tests/phpunit/tests/kses.php

```diff
@@ -1357,15 +1357,139 @@ public function data_safecss_filter_attr() {
 	}
 
 	/**
-	 * Data attributes are globally accepted.
+	 * Ensures proper recognition of a data attribute and how to transform its
+	 * name into what JavaScript code would read from an element's `dataset`.
+	 *
+	 * @ticket 61052
+	 *
+	 * @dataProvider data_possible_custom_data_attributes_and_transformed_names
+	 *
+	 * @param string      $attribute_name      Raw attribute name.
+	 * @param string|null $dataset_name_if_any Transformed attribute name, or `null`
+	 *                                         if not a custom data attribute.
+	 */
+	public function test_wp_kses_transform_custom_data_attribute_name_recognizes_data_attributes( $attribute_name, $dataset_name_if_any ) {
+		$transformed_name = wp_kses_transform_custom_data_attribute_name( $attribute_name );
+
+		if ( isset( $dataset_name_if_any ) ) {
+			$this->assertNotNull(
+				$transformed_name,
+				"Failed to recognize '{$attribute_name}' as a custom data attribute."
+			);
+
+			$this->assertSame(
+				$dataset_name_if_any,
+				$transformed_name,
+				'Improperly transformed custom data attribute name.'
+			);
+		} else {
+			$this->assertNull(
+				$transformed_name,
+				"Should not have identified '{$attribute_name}' as a custom data attribute."
+			);
+		}
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array[].
+	 */
+	public static function data_possible_custom_data_attributes_and_transformed_names() {
+		return array(
+			// Non-custom-data attributes.
+			'Normal attribute'             => array( 'post-id', null ),
+			'Single word'                  => array( 'id', null ),
+
+			// Normative custom data attributes.
+			'Normal custom data attribute' => array( 'data-post-id', 'postId' ),
+			'Leading dash'                 => array( 'data--before', 'Before' ),
+			'Trailing dash'                => array( 'data-after-', 'after-' ),
+			'Double-dashes'                => array( 'data-wp-bind--enabled', 'wpBind-Enabled' ),
+			'Double-dashes everywhere'     => array( 'data--one--two--', 'One-Two--' ),
+			'Triple-dashes'                => array( 'data---one---two---', '-One--Two---' ),
+
+			// Unexpected but recognized custom data attributes.
+			'Only comprising a prefix'     => array( 'data-', '' ),
+			'With upper case ASCII'        => array( 'data-Post-ID', 'postId' ),
+			'With medial upper casing'     => array( 'data-uPPer-cAsE', 'upperCase' ),
+			'With Unicode whitespace'      => array( "data-\u{2003}", "\u{2003}" ),
+			'With Emoji'                   => array( 'data-🐄-pasture', '🐄Pasture' ),
+			'Brackets and colon'           => array( 'data-[wish:granted]', '[wish:granted]' ),
```

Comment:
I confirmed that (surprisingly) all these cases do seem to work with way in the browser.

Reply:
Thanks! I didn't know how to automate them in WordPress' tests, but I don't know if you saw that I included a test script in the PR description which I used to confirm each case. In PHP I use the old detector in WordPress, the one in this patch, and then in JavaScript I set the `innerText` to be the property name(s) of the available properties in an element's own `dataset`, which shows how the browser transforms them.

The included screenshot is from the output of this script, and I compared all three big browsers.

Reply:
I missed the test script! Thanks for providing it 👍

I did something very similar myself. On my first attempt, I tried using DOM APIs to create an element and set the attribute. _That_ (`setAttribute`) rejects some attribute names that worked fine when setting the data attributes in HTML.